### ISSUE TITLE
API: Expand read_csv dtype for categoricals

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#14503]
In https://github.com/pandas-dev/pandas/pull/13406 Chris added support for `read_csv(..., dtype={'col': 'category'})` (thanks!). This issue is for expanding that syntax to allow a more complete specification of the resulting categorical.

``` python
# Your code here
df = pd.read_csv(path, dtype={'col': pd.Categorical(['a', 'b', 'c'], ordered=True})
df = pd.read_csv(path, dtype={'col': ['a', 'b', 'c']})  # shorthand, but unordered only
# we would still accept `dtype={'col': 'category'}` as well, to infer categories
```

Implementation-wise, I think we can keep all the parsing logic as is, and simply loop over `dtype` and call `set_categories` (and maybe `as_ordered`) on all the categoricals just before returning to the user.

This would help a bit in dask, where their category type inference can fail if the first partition doesn't contain all the categories (see https://github.com/dask/dask/issues/1705). This is why it'd be preferable to do it as an option to `read_csv`, rather than putting in on the user to followup with a `set_categories`.

**Comment by chris-b1** [comment]
If it matters, there would be at least a little performance to be picked up by modifying the parsing code - you could pass the categories in here:

https://github.com/pandas-dev/pandas/blob/7a2bcb6605bacea858ec14cfac424898deb568b3/pandas/parser.pyx#L1523

And shortcut building the categories array, sorting, etc.  It would also cause an error to be thrown much earlier if the data has a value not in the specified categories.

**Comment by TomAugspurger** [comment]
Yes, I think you're right that it'd be better to do it in the Cython code.
I was digging through there last week, and it shouldn't be to much extra
effort.

Though I guess we'll need to have an API discussion about what to do if the
user passes `dtype={'A': pd.Categorical(['a', 'b'])}` and a value outside
those shows up.

1. Throw an exception
2. Set to `NA`

The second option would be consistent with `set_categories`.

```python
In [1]: pd.Categorical(['a', 'b', 'c']).set_categories(['a', 'b'])
Out[1]:
[a, b, NaN]
Categories (2, object): [a, b]
```

**Comment by lawriewelsby** [comment]
Hi, I have tried to follow the chain of issues that branches off from this but can't seem to find out if this change has been implemented (appears not though the work to make it possible has been done?) and if so what the syntax would be. Any update would be much appreciated. Many thanks.

**Comment by jbrockmendel** [comment]
users can specify `CategoricalDtype(...)`.  I'd rather not complicate the API to save a few keystrokes.

**Comment by TomAugspurger** [comment]
Agreed.
